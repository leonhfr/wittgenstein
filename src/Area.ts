// Internal.
import * as Types from './types';

// Code.
export class Area {
  readonly id: string;
  readonly name: string;
  readonly file: string;
  readonly lastScheduledAt: number;
  readonly refreshRate: number;
  readonly enabled: boolean;
  readonly zonesComputed: boolean;

  static create(input: unknown): Area | Error {
    if (isArea(input)) {
      return new Area(input);
    }
    const errMsg = isSafeArea(input).errMsg;
    return new Error(`ValidationError: ${errMsg}`);
  }

  private constructor(input: CreateAreaInput) {
    this.id = input.id;
    this.name = input.name;
    this.file = input.file;
    this.lastScheduledAt = input.lastScheduledAt;
    this.refreshRate = input.refreshRate;
    this.enabled = input.enabled;
    this.zonesComputed = input.zonesComputed;
  }
}

// Validation.

export const AREA_PROPS = [
  'id',
  'name',
  'file',
  'lastScheduledAt',
  'refreshRate',
  'enabled',
  'zonesComputed',
];

export const isArea = (input: unknown): input is CreateAreaInput => {
  return isSafeArea(input).isSafe;
};

export const isSafeArea = (input: unknown): Types.IsSafe => {
  if (typeof input !== 'object') {
    return {
      isSafe: false,
      errMsg: `Expected type of input to be object, got ${typeof input}.`,
    };
  }

  if (input === null) {
    return {
      isSafe: false,
      errMsg: `Expected input not to be null.`,
    };
  }

  for (const key of AREA_PROPS) {
    if (!input.hasOwnProperty(key)) {
      return {
        isSafe: false,
        errMsg: `Expected all input properties to be set, missing ${key}.`,
      };
    }
  }

  const {
    id,
    name,
    file,
    lastScheduledAt,
    refreshRate,
    enabled,
    zonesComputed,
  } = input as {
    id: unknown;
    name: unknown;
    file: unknown;
    lastScheduledAt: unknown;
    refreshRate: unknown;
    enabled: unknown;
    zonesComputed: unknown;
  };

  const numbers = { lastScheduledAt, refreshRate };

  const strings = { id, name, file };

  const booleans = { enabled, zonesComputed };

  // Numbers

  for (const [key, value] of Object.entries(numbers)) {
    if (typeof value !== 'number') {
      return {
        isSafe: false,
        errMsg: `Expected type of input.${key} to be number, got ${typeof value}.`,
      };
    }
  }

  // Strings

  for (const [key, value] of Object.entries(strings)) {
    if (typeof value !== 'string') {
      return {
        isSafe: false,
        errMsg: `Expected type of input.${key} to be string, got ${typeof value}.`,
      };
    }
  }

  // Booleans

  for (const [key, value] of Object.entries(booleans)) {
    if (typeof value !== 'boolean') {
      return {
        isSafe: false,
        errMsg: `Expected type of input.${key} to be boolean, got ${typeof value}.`,
      };
    }
  }

  return {
    isSafe: true,
    errMsg: '',
  };
};

// Interface

export interface CreateAreaInput {
  readonly id: string;
  readonly name: string;
  readonly file: string;
  readonly lastScheduledAt: number;
  readonly refreshRate: number;
  readonly enabled: boolean;
  readonly zonesComputed: boolean;
}
