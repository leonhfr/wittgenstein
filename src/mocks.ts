// Internal.
import { CreateAreaInput } from './Area';
import { CreatePhotoInput } from './Photo';

// Code.
export const createMockAreaInput = (area?: any): CreateAreaInput => {
  const _area = area || {};
  return {
    id: _area.id || '8ea8f412-4ee8-4f1d-9228-4c133b8fbd37',
    name: _area.name || 'Sant Marti',
    file: _area.file || 'sant-marti',
    lastScheduledAt: _area.lastScheduledAt || 0,
    refreshRate: _area.refreshRate || 86400,
    enabled: _area.enabled || true,
    zonesComputed: _area.zonesComputed || false,
  };
};

export const createMockPhotoInput = (photo?: any): CreatePhotoInput => {
  const _photo = photo || {};
  return {
    id: _photo.id || '32457360587',
    owner: _photo.owner || '41710899@N08',
    secret: _photo.secret || '69b27ceb18',
    server: _photo.server || '4898',
    farm: _photo.farm || 5,
    title: _photo.title || 'Arc de Triomf',
    description:
      _photo.description ||
      'The Arc de Triomf was built as the entrance of the 1888 Barcelona Universal Exposition. It was designed by architect Josep Vilaseca i Casanovas in the Neo-Mudéjar style.',
    ownername: _photo.ownername || 'jdf_92',
    views: _photo.views || 16,
    tags: _photo.tags || [
      'spain',
      'barcelona',
      'arcdetriomf',
      '1888barcelonaworldfair',
      'neomudéjar',
      '1888barcelonauniversalexposition',
      'catelonia',
    ],
    latitude: _photo.latitude || 41.390791,
    longitude: _photo.longitude || 2.18114,
    context: _photo.context || 0,
  };
};
