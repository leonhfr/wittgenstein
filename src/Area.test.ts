// Internal.
import { Area, AREA_PROPS, isArea } from './Area';
import * as Mocks from './mocks';

// Code.
describe('Area', () => {
  describe('Class', () => {
    it('should be defined', () => {
      expect(Area).toBeDefined();
    });

    describe('create', () => {
      it('should fail to create a new instance - invalid input', () => {
        expect(Area.create(null)).toMatchSnapshot();
      });
      it('should create a new instance', () => {
        expect(Area.create(Mocks.createMockAreaInput())).toMatchSnapshot();
      });
    });
  });

  describe('Validation', () => {
    it('should be defined', () => {
      expect(AREA_PROPS).toBeDefined();
    });
    it('should match the expected value', () => {
      expect(AREA_PROPS).toMatchSnapshot();
    });
    it('should be up to date', () => {
      expect(AREA_PROPS).toEqual(
        Object.keys(Area.create(Mocks.createMockAreaInput()))
      );
    });
  });

  describe('isArea', () => {
    it('should be defined', () => {
      expect(isArea).toBeDefined();
    });
    it('should handle an invalid input - wrong type', () => {
      expect(isArea('__FAIL__')).toBe(false);
    });
    it('should handle an invalid input - null type', () => {
      expect(isArea(null)).toBe(false);
    });
    it('should handle an invalid input - missing props', () => {
      expect(isArea({})).toBe(false);
    });
    it('should handle an invalid input - numbers', () => {
      expect(
        isArea({
          ...Mocks.createMockAreaInput(),
          refreshRate: '__FAIL__',
        })
      ).toBe(false);
    });
    it('should handle an invalid input - strings', () => {
      expect(
        isArea({
          ...Mocks.createMockAreaInput(),
          id: false,
        })
      ).toBe(false);
    });
    it('should handle an invalid input - booleans', () => {
      expect(
        isArea({
          ...Mocks.createMockAreaInput(),
          enabled: '__FAIL__',
        })
      ).toBe(false);
    });
  });
});
