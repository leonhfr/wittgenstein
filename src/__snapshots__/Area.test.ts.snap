// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Area Class create should create a new instance 1`] = `
Area {
  "enabled": true,
  "file": "sant-marti",
  "id": "8ea8f412-4ee8-4f1d-9228-4c133b8fbd37",
  "lastScheduledAt": 0,
  "name": "Sant Marti",
  "refreshRate": 86400,
  "zonesComputed": false,
}
`;

exports[`Area Class create should fail to create a new instance - invalid input 1`] = `[Error: ValidationError: Expected input not to be null.]`;

exports[`Area Validation should match the expected value 1`] = `
Array [
  "id",
  "name",
  "file",
  "lastScheduledAt",
  "refreshRate",
  "enabled",
  "zonesComputed",
]
`;
